otherlibs/unix/link.c

```diff
@@ -15,7 +15,9 @@
 
 /* Needed to get linkat exposed in compliant OS.
    Must be defined before the first system .h is included. */
+#if !defined(__SUNPRO_C)
 #define _XOPEN_SOURCE 700
+#endif
```

Comment:
I need to understand the problem better.  `_XOPEN_SOURCE 700` corresponds to the Single Unix Specification version 4.  https://docs.oracle.com/cd/E88353_01/html/E37853/xpg6-7.html says that SuSv4 is supported provided the `c99` option is given to the compiler, and then `_XOPEN_SOURCE` is set to 700 (so, re-defining it should be a no-op).

1. Why are things not working for you as described in Oracle's docs?
2. If you compile with gcc but with the same set of standard include files, `__SUNPRO_C` will be false and `_XOPEN_SOURCE` will be defined to `700`.  Does this cause the same errors?
3.

Reply:
On 

`SunOS sundev13 5.11 11.3 sun4v sparc sun4v Solaris` with `Sun PRO C` I get an error

```bash
"/usr/include/sys/feature_tests.h", line 354: #error: "Compiler or options invalid for pre-UNIX 03 X/Open applications 	and pre-2001 POSIX applications"
```

another way to fix it is to define `-D_XPG6=""` for `Solaris/Sun PRO C`, which can be done in `configure`! Thanks a lot, @xavierleroy , we need to remove this change to link.c!

So,

1. Perhaps it is because `_XPG6` was not defined. Oracle docs say (see your link)

_"A default Oracle Solaris installation might require additional steps in order to be fully SUSv4 compliant."_

apparently we do not have these steps fulfilled. The best workaround seems to define `_XPG6`.

2. Unfortunately we do not have `gcc` on `SPARC` systems in BB. I tried `clang`, and yes, it does not like this place (float.c and link.c have the same `_XOPEN_SOURCE` define):

```bash
clang -c -O2 -fno-strict-aliasing -fwrapv -Wall -Wdeclaration-after-statement -Werror -fno-common -g  -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE   -DCAMLDLLIMPORT=  -std=c99 -o floats.b.o floats.c
floats.c:21:9: error: '_XOPEN_SOURCE' macro redefined [-Werror,-Wmacro-redefined]
#define _XOPEN_SOURCE 700
        ^
<built-in>:311:9: note: previous definition is here
#define _XOPEN_SOURCE 600
        ^
In file included from floats.c:26:
In file included from /usr/include/math.h:17:
In file included from /usr/include/iso/math_iso.h:12:
/usr/include/sys/feature_tests.h:354:2: error: "Compiler or options invalid for pre-UNIX 03 X/Open applications 	and pre-2001 POSIX applications"
#error "Compiler or options invalid for pre-UNIX 03 X/Open applications \
 ^
2 errors generated.
```

defining `_XPG6` solves only the second error:

```bash
clang -c -O2 -fno-strict-aliasing -fwrapv -Wall -Wdeclaration-after-statement -Werror -fno-common -g  -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE   -DCAMLDLLIMPORT=  -D_XPG6="" -std=c99 -o floats.b.o floats.c
floats.c:21:9: error: '_XOPEN_SOURCE' macro redefined [-Werror,-Wmacro-redefined]
#define _XOPEN_SOURCE 700
        ^
<built-in>:311:9: note: previous definition is here
#define _XOPEN_SOURCE 600
        ^
1 error generated.
```